src/lib/fixer.ts

```diff
@@ -131,6 +138,164 @@ export class _PmdFixGenerator extends FixGenerator {
+            }
+        } else {
+            // If @SuppressWarnings does not exist, insert a new one
+            const newSuppression = this.generateNewSuppressionTag(suppressionRule, this.document.languageId);
+            action.edit.insert(this.document.uri, classStartPosition, newSuppression);
+        }
+    
+        action.diagnostics = [this.diagnostic];
+        action.command = {
+            command: Constants.COMMAND_REMOVE_DIAGNOSTICS_ON_SELECTED_FILE,
+            title: 'Remove diagnostics for this file',
+            arguments: [this.document.uri]
+        };
+
+        return action;
+    }
+    
+    public generateUpdatedSuppressionTag(updatedRules: string, lang: string) {
+        if (lang === 'apex') {
+            return `@SuppressWarnings('${updatedRules}')`;
+        } else if (lang === 'java') {
+            return `@SuppressWarnings("${updatedRules}")`;
+        }
+        return '';
+    }
+
+    public generateNewSuppressionTag(suppressionRule: string, lang: string) {
+        if (lang === 'apex') {
+            return `@SuppressWarnings('${suppressionRule}')\n`;
+        } else if (lang === 'java') {
+            return `@SuppressWarnings("${suppressionRule}")\n`;
+        }
+        return '';
+    }
+
+    /**
+     * Finds the start position of the class in the document.
+     * Assumes that the class declaration starts with the keyword "class".
+     * @returns The position at the start of the class.
+     */
+    public findClassStartPosition(diagnostic: vscode.Diagnostic, document: vscode.TextDocument): vscode.Position {
+        const text = document.getText();
+        const diagnosticLine = diagnostic.range.start.line;
+    
+        // Split the text into lines for easier processing
+        const lines = text.split('\n');
+        let classStartLine: number | undefined;
+
+        let inBlockComment = false;
+    
+        // Iterate from the diagnostic line upwards to find the class declaration
+        for (let lineNumber = diagnosticLine; lineNumber >= 0; lineNumber--) {
+            const line = lines[lineNumber];
+
+            // Check if we are in a block comment
+            if (inBlockComment) {
+                if (line.match(this.blockCommentStartPattern)) {
+                    inBlockComment = false;
+                }
+                continue;
+            }
+
+            // Skip single-line comments
+            if (line.match(this.singleLineCommentPattern)) {
+                continue;
+            }
+
+            // Skip block comment in a single line
+            if (line.match(this.blockCommentEndPattern) && line.match(this.blockCommentStartPattern)) {
+                continue;
+            }
+
+            // Check if this line is the start of a block comment
+            if (line.match(this.blockCommentEndPattern)) {
+                inBlockComment = true;
+                continue;
+            }
```

Comment:
I'm a bit concerned how strings are going to prevent these regular expressions from working as you desire them to. For example, if I have a string abc = 'some */ random string'; then it will think you just entered into a block comment (specifically because you are going backward). If you were going forward, the blockCommentStartPattern is more robust.

So I propose you first walk forward through the file and keep track of what lines are comments (since deteching comments blocks are easier going forward). Then afterwards walk backward from the diagnostic line and skip those commented lines. 

Alternatively maybe your isWithinQuotes will help here.

Reply:
I agree with Stephen. I'd add tests for this kind of corner case if you can.

Reply:
Took care of this.